### ISSUE TITLE
Missing titles for properties and enums in Service Plans' JSON schema

### ISSUE THREAD

**Issue by pkosiec** [Azure/open-service-broker-azure#651]
**Description**

In some JSON schemas for Service Plans, there are properties without `title` property. Also, in enums there are no labels for values, which can be achieved in JSON schema with `oneOf`.

**Reasons**

The `title` property is widely used by JSON schema form rendering libraries as a field label. The suggested change, as well as introducing friendly, human-readable labels for enums, will greatly improve usability on UI. 

It's very similar to issue #519. 

**Examples** 

- Missing title in `alias` property for `Azure Database for MySQL 5.7-- DBMS Only` plans
- Missing title for `parentAlias` property:

![screen shot 2018-12-14 at 11 14 02](https://user-images.githubusercontent.com/7155799/49998224-0477c180-ff94-11e8-9773-1b937a0b4c7d.png)

- Missing enum value titles for "Backup redundancy" enum:

![screen shot 2018-12-14 at 11 18 48](https://user-images.githubusercontent.com/7155799/49998216-004ba400-ff94-11e8-8d1e-7452540f4d45.png)

---

**Expected result**
This is how the form based on JSON schema renders when all titles have `title` specified (as an example: Azure SQL Database 12.0 plans):

![screen shot 2018-12-14 at 11 13 41](https://user-images.githubusercontent.com/7155799/49998247-15283780-ff94-11e8-9ece-ffb15d5ab3d0.png)

**Comment by pkosiec** [comment]
@jldeen please take a look at this. Thank you!

**Comment by norshtein** [comment]
Hi @pkosiec , thanks for raising that. The first and the second item will be fixed soon. For the third item 'Missing enum value titles for "Backup redundancy" enum', `enum` is a json array in OSBA's [implementation](https://github.com/Azure/open-service-broker-azure/blob/master/pkg/service/schema.go#L111), each element in this array must be a string, and can't be an object, so a title can't be added to the enum value.

**Comment by pkosiec** [comment]
Hi @norshtein,
Thank you for very fast reaction and your PR! I hope that all titles for all fields are covered now 🙂 

When it comes to the labels for enum values: yes, it would result in slightly changing the implementation. But not much - user would still provide the same `instanceCreateParameterSchema`. The definition of the JSON schema for Service Plans would change in the following way:

So, instead of having this specification:
```
{
    "title": "Before",
    "type": "object",
    "required": [
      "firstName",
      "lastName"
    ],
    "properties": {
        "backupRedundancy": {
            "default": "local",
            "description": "Specifies the backup redundancy",
            "enum": ["local", "geo"],
            "title": "Backup redundancy",
            "type": "string"
          }
    }
}
```

We could have:

```
{
    "title": "After",
    "type": "object",
    "properties": {
      "backupRedundancy": {
        "default": "geo",
        "description": "Specifies the backup redundancy",
        "oneOf": [
          {
            "enum": ["local"],
            "title": "Local Label"
          },
          {
            "enum": ["geo"],
            "title": "Geo Label"
          }
        ],
        "title": "Backup redundancy",
        "type": "string"
      }
    }
  }
```

So I guess internal types and the implementation of converting provided schema would change and that's all.

**Comment by norshtein** [comment]
Hi @pkosiec, is the behavior defined by any standard? It will make us in a dilemma if every organization proposes a different implementation... Or you may generate a label at frontend layer for every element in `enum` array.

**Comment by pkosiec** [comment]
Hello @norshtein,

This is 100% valid JSON schema. And this is currently the only way how to define labels for enum values - see https://github.com/json-schema-org/json-schema-spec/issues/57.

I've checked quickly front-end libraries and I tested front-end libraries for Angular & React, which support `oneOf` rendering as `enum` with nice labels:
https://github.com/hamzahamidi/angular6-json-schema-form 
https://github.com/mozilla-services/react-jsonschema-form

Generating labels on front-end side is not a best option, as for example converting `geo` to `Geo` still doesn't say much to the UI user.

**Comment by pkosiec** [comment]
What do you mean by another case of enum? The #655 issue is about converting string property to enum one. I didn't want to put two ideas in one issue, so I gave there (in #655) an example how the enum would look like after conversion right now - to be 100% consistent with others enums you already have.

But, here, in this issue, the idea is to replace every enum with `anyOf, as in the example given above. So yes, the `location` enum would be replaced with `anyOf` as well and there would be no inconsistency.

**Comment by zhongyi-zhang** [comment]
It makes sense to me. Reopening the issue.
But we have to work on some other very high priority items now. So it might not be addressed in short time :(. And of course if you have interest in it, your contribution would be super appreciated!

**Comment by piotrmiskiewicz** [comment]
I've prepared a PR which introduced the `oneOf` field in schemas: https://github.com/Azure/open-service-broker-azure/pull/665

**Comment by zhongyi-zhang** [comment]
Closed via #665.
